processing/src/main/java/org/apache/druid/segment/nested/NestedFieldTypeInfo.java

```diff
@@ -31,29 +32,36 @@
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.channels.WritableByteChannel;
+import java.util.Objects;
 import java.util.Set;
 
 /**
  * Binary serialization for nested field type info, translated into this compact format for storage in segments.
  * The index of the type info here is the same as the field index in {@link CompressedNestedDataComplexColumn#fields}
  */
-public class NestedLiteralTypeInfo
+public class NestedFieldTypeInfo
 {
   private static final byte STRING_MASK = 1;
   private static final byte LONG_MASK = 1 << 2;
   private static final byte DOUBLE_MASK = 1 << 3;
 
-  public static NestedLiteralTypeInfo read(ByteBuffer buffer, int length)
+  private static final byte STRING_ARRAY_MASK = 1 << 4;
+
+  private static final byte LONG_ARRAY_MASK = 1 << 5;
+
+  private static final byte DOUBLE_ARRAY_MASK = 1 << 6;
```

Comment:
I dunno if I'm being stingy, but couldn't we have an `array` mask that is applied and then we check the type mask after that?  Would mean that we can add array support to this and only consume 1 more bit instead of consuming 3 more bits.

Reply:
that would work if we only set 1 type, or if any row being an array meant all rows are arrays, but it doesn't currently, so we sort of need to know what type of array, at least the way things currently work

Reply:
what about composite arrays?  Like if the array has a String and a double?  Or a String and an Object?

I had intepretted these as "this type exists in the column" not as "every row is this type".  In which case, if there is an array, then it is true that the array type exists in the column, no?  And then if the only scalar type is a `long`, then you would know that the arrays are all longs, right?